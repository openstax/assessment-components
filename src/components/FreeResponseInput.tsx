import { MouseEventHandler, ReactNode } from 'react';
import { countWords } from '../utils';
import styled, { css } from 'styled-components';
import { colors, mixins } from '../theme';
import { AvailablePoints, ExerciseQuestionData } from 'src/types';
import { QuestionHtml } from './Question';
import Button from './Button';
import { StepCardFooter } from './StepCardFooter';

export interface FreeResponseProps {
  readOnly: boolean;
  wordLimit: number;
  infoRowChildren?: ReactNode;
  onChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
  cancelHandler: MouseEventHandler<HTMLButtonElement>;
  saveHandler: MouseEventHandler<HTMLButtonElement>;
  defaultValue: string;
  isSubmitDisabled: boolean;
  questionNumber: number,
  question: ExerciseQuestionData,
  availablePoints?: AvailablePoints,
  textHasChanged: boolean;
  submitBtnLabel: string;
}

const TextAreaErrorStyle = css`
  background-color: #f5e9ea;
`;

const StyledFreeResponse = styled.div`
  display: flex;
  flex-direction: column;

  .step-card-body {
    ${mixins.stepCardPadding()}
  }
`;

const SyledQuestionStem = styled.div`
  font-size: 2rem;
  line-height: 1.68em;
  position: relative;
`;

const InfoRow = styled.div<{ hasChildren: boolean }>`
  margin: 8px 0;
  display: flex;
  justify-content: ${props => props.hasChildren ? 'space-between' : 'flex-end'};
  line-height: 1.6rem;

  .word-limit-error-info {
    color: ${colors.palette.danger};
  }

  div > span {
    font-size: 12px;
    line-height: 16px;

    + span {
      margin-left: 1rem;
    }
  }

  .last-submitted + * {
    margin-top: 0.8rem;
  }

  color: ${colors.palette.neutralThin};
`;

export const FreeResponseTextArea = styled.textarea<{ isOverWordLimit: boolean } & FreeResponseProps>`
  display: block;
  font-family: inherit;
  font-size: 1.8rem;
  line-height: 3rem;
  width: 100%;
  min-height: 10.5em;
  line-height: 1.5em;
  margin: 2.5rem 0 0 0;
  padding: 0.5em;
  border: 1px solid ${colors.palette.neutral};
  color: ${colors.palette.neutralDark};
  ${props => props.isOverWordLimit && TextAreaErrorStyle};
  ${props => props.isOverWordLimit && css`
    border: 2px solid ${colors.palette.danger};
  `}
  background-color: ${props => props.readOnly && colors.palette.neutralCool};
`;
FreeResponseTextArea.displayName = 'OSFreeResponseTextArea';

const RevertButton = (props: {
  disabled: boolean
} & React.ComponentPropsWithoutRef<'button'>) => (
  <Button {...props}>
    Cancel
  </Button>
);


export const FreeResponseInput = (props: FreeResponseProps) => {
  const {
    availablePoints,
    cancelHandler,
    defaultValue,
    infoRowChildren,
    isSubmitDisabled,
    question,
    questionNumber,
    saveHandler,
    submitBtnLabel,
    textHasChanged,
    wordLimit,
  } = props;

  const isOverWordLimit = countWords(defaultValue) > wordLimit;

  const questionProps = {};
  if (questionNumber) { questionProps['data-question-number'] = questionNumber; }

  return (
    <StyledFreeResponse data-test-id="student-free-response">
      <div className="step-card-body">
        <SyledQuestionStem {...questionProps}>
          {question.stem_html &&
            <QuestionHtml type="stem" html={question.stem_html} hidden={false} />}
        </SyledQuestionStem>
        <FreeResponseTextArea
          {...props}
          isOverWordLimit={isOverWordLimit}
          data-test-id="free-response-box"
          placeholder="Enter your response..."
          aria-label="question response text box"
        />
        <InfoRow hasChildren={!!infoRowChildren}>
          {infoRowChildren}
          <div>
            <span>{countWords(defaultValue)} words</span>
            {isOverWordLimit && <span className="word-limit-error-info">Maximum {wordLimit} words</span>}
          </div>
        </InfoRow>
      </div>
      <StepCardFooter>
      {availablePoints
        ? <div className="points"><strong>Points: {availablePoints}</strong></div>
        : null}
        <div className="controls">
          <RevertButton disabled={!textHasChanged} onClick={cancelHandler} />
          <Button
            data-test-id="submit-answer-btn"
            disabled={isSubmitDisabled || isOverWordLimit}
            onClick={saveHandler}
          >
            {submitBtnLabel}
          </Button>
        </div>
      </StepCardFooter>
    </StyledFreeResponse>
  );
}

FreeResponseInput.displayName = 'OSFreeResponse';
