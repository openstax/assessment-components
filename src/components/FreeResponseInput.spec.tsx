import { FreeResponseInput, FreeResponseProps } from './FreeResponseInput';
import renderer from 'react-test-renderer';
import React from 'react';

describe('Free Response Input', () => {
  let props: FreeResponseProps;

  const updateValue = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    props.defaultValue = event.target.value
  };

  beforeEach(() => {
    props = {
      isErrored: false,
      isDisplayingNudge: false,
      readOnly: false,
      lastSubmitted: '',
      wordLimit: 5,
      submittedComponent: <span className="last-submitted">Last submitted on July 26 at 4:00 pm</span>,
      onChange: updateValue,
      defaultValue: '',
    };
  });

  it('matches snapshot', () => {
     const tree = renderer.create(
       <FreeResponseInput {...props} readOnly={true} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders last submitted text', () => {
    const tree = renderer.create(
      <FreeResponseInput {...props} lastSubmitted={'2015-10-06T11:59:00.000Z'} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders word limit error', () => {
    const tree = renderer.create(
      <FreeResponseInput {...props} isDisplayingNudge={true} isErrored={true} defaultValue={'response has more than five words'} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
