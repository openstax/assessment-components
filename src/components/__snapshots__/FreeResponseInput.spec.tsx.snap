// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Free Response Input matches snapshot 1`] = `
<div
  className="sc-bczRLJ goEvZo"
  data-test-id="student-free-response"
>
  <div
    className="step-card-body"
  >
    <textarea
      aria-label="question response text box"
      className="sc-dkzDqf brPzWs"
      data-test-id="free-response-box"
      defaultValue=""
      onChange={[Function]}
      placeholder="Enter your response..."
      readOnly={true}
    />
    <div
      className="sc-gsnTZi RJiQN"
    >
      <div />
      <div>
        <span>
          0
           words
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Free Response Input renders last submitted text 1`] = `
<div
  className="sc-bczRLJ goEvZo"
  data-test-id="student-free-response"
>
  <div
    className="step-card-body"
  >
    <textarea
      aria-label="question response text box"
      className="sc-dkzDqf loLWrQ"
      data-test-id="free-response-box"
      defaultValue=""
      onChange={[Function]}
      placeholder="Enter your response..."
      readOnly={false}
    />
    <div
      className="sc-gsnTZi RJiQN"
    >
      <div>
        <span
          className="last-submitted"
        >
          Last submitted on July 26 at 4:00 pm
        </span>
      </div>
      <div>
        <span>
          0
           words
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Free Response Input renders word limit error 1`] = `
<div
  className="sc-bczRLJ goEvZo"
  data-test-id="student-free-response"
>
  <div
    className="step-card-body"
  >
    <textarea
      aria-label="question response text box"
      className="sc-dkzDqf iLsrhj"
      data-test-id="free-response-box"
      defaultValue="response has more than five words"
      onChange={[Function]}
      placeholder="Enter your response..."
      readOnly={false}
    />
    <div
      className="sc-gsnTZi RJiQN"
    >
      <div />
      <div>
        <span>
          6
           words
        </span>
        <span
          className="word-limit-error-info"
        >
          Maximum 
          5
           words
        </span>
      </div>
    </div>
  </div>
</div>
`;
